From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: peaches94 <peachescu94@gmail.com>
Date: Sat, 2 Jul 2022 00:35:56 -0500
Subject: [PATCH] Multithreaded Tracker

Original license: GPL-3.0-only
Original project: https://github.com/Bloom-host/Petal

Original license: GPL-3.0-only
Original project: https://github.com/TECHNOVE/Airplane-Experimental

Co-authored-by: Paul Sauve <paul@technove.co>
Co-authored-by: Kevin Raneri <kevin.raneri@gmail.com>
Co-authored-by: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Co-authored-by: hayanesuru <hayanesuru@outlook.jp>

This patch refactored from original multithreaded tracker (Petal version),
and is derived from the Airplane fork by Paul Sauve, the tree is like:
Airplane -> Pufferfish(?) -> Petal -> Leaf

The core logic has beed reworked compared to the old one, can handle larger
scale situation better now.
Current impl includes many improvements and fixes we made, such as
plugin compat issues with some NPC plugins using real entity type,
e.g. Citizens.

However we still recommend to use those packet based NPC plugins,
e.g. ZNPC Plus, Adyeshach, Fancy NPC, etc.

diff --git a/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java b/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
index e227d101dec76100be6c24db9d4a8c5cb7028128..82b3a42b2d7d9d027596b00af328e060843c87c9 100644
--- a/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
+++ b/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
@@ -164,6 +164,8 @@ public final class NearbyPlayers {
         private int nonEmptyLists;
         private long updateCount;
 
+        public final it.unimi.dsi.fastutil.objects.ReferenceSet<ServerPlayer> playersTracking = org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled ? it.unimi.dsi.fastutil.objects.ReferenceSets.synchronize(new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(1)) : null; // Leaf - Multithreaded tracker
+        public final java.util.concurrent.atomic.AtomicLong trackingUpdateCountAtomic = org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled ? new java.util.concurrent.atomic.AtomicLong(0) : null; // Leaf - Multithreaded tracker
         public TrackedChunk(final long chunkKey, final NearbyPlayers nearbyPlayers) {
             this.chunkKey = chunkKey;
             this.nearbyPlayers = nearbyPlayers;
@@ -177,6 +179,12 @@ public final class NearbyPlayers {
             return this.updateCount;
         }
 
+        // Leaf start - Multithreaded tracker
+        public long getAtomicUpdateCount() {
+            return this.trackingUpdateCountAtomic.get() | 0x1000000000000000L;
+        }
+        // Leaf end - Multithreaded tracker
+
         public ReferenceList<ServerPlayer> getPlayers(final NearbyMapType type) {
             return this.players[type.ordinal()];
         }
@@ -185,6 +193,12 @@ public final class NearbyPlayers {
             ++this.updateCount;
 
             final int idx = type.ordinal();
+            // Leaf start - Multithreaded tracker
+            if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled && idx == NearbyMapType.VIEW_DISTANCE.ordinal()) {
+                this.trackingUpdateCountAtomic.getAndIncrement();
+                this.playersTracking.add(player);
+            }
+            // Leaf end - Multithreaded tracker
             final ReferenceList<ServerPlayer> list = this.players[idx];
             if (list == null) {
                 ++this.nonEmptyLists;
@@ -203,6 +217,12 @@ public final class NearbyPlayers {
             ++this.updateCount;
 
             final int idx = type.ordinal();
+            // Leaf start - Multithreaded tracker
+            if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled && type == NearbyMapType.VIEW_DISTANCE) {
+                this.trackingUpdateCountAtomic.getAndIncrement();
+                this.playersTracking.remove(player);
+            }
+            // Leaf end - Multithreaded tracker
             final ReferenceList<ServerPlayer> list = this.players[idx];
             if (list == null) {
                 throw new IllegalStateException("Does not contain player " + player);
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 8b987ca4e1db327f42806083c25e758375a959ef..bada1d995614d028160166002555d261821ad7dc 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -345,7 +345,7 @@ public final class RegionizedPlayerChunkLoader {
         private boolean canGenerateChunks = true;
 
         private final ArrayDeque<ChunkHolderManager.TicketOperation<?, ?>> delayedTicketOps = new ArrayDeque<>();
-        private final LongOpenHashSet sentChunks = new LongOpenHashSet();
+        private final LongOpenHashSet sentChunks = org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled ? new org.dreeam.leaf.util.map.SyncLongOpenHashSet() : new LongOpenHashSet(); // Leaf - Multithreaded tracker
 
         private static final byte CHUNK_TICKET_STAGE_NONE           = 0;
         private static final byte CHUNK_TICKET_STAGE_LOADING        = 1;
@@ -441,6 +441,14 @@ public final class RegionizedPlayerChunkLoader {
 
                 PlatformHooks.get().onChunkWatch(this.world, chunk, this.player);
                 PlayerChunkSender.sendChunk(this.player.connection, this.world, chunk);
+                // Leaf start - Multithreaded tracker
+                if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) {
+                    ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk trackedChunk = chunk.moonrise$getChunkHolder().holderData.nearbyPlayers;
+                    if (trackedChunk != null) {
+                        trackedChunk.trackingUpdateCountAtomic.getAndIncrement();
+                    }
+                }
+                // Leaf end - Multithreaded tracker
                 return;
             }
             throw new IllegalStateException();
diff --git a/io/papermc/paper/FeatureHooks.java b/io/papermc/paper/FeatureHooks.java
index 504fdb2c4000eafae4341ec8b24ce9cb0fcdc274..d600bd9f0b26005966eaba943fc38012c96599ea 100644
--- a/io/papermc/paper/FeatureHooks.java
+++ b/io/papermc/paper/FeatureHooks.java
@@ -69,6 +69,14 @@ public final class FeatureHooks {
     }
 
     public static Set<Long> getSentChunkKeys(final ServerPlayer player) {
+        // Leaf start - Multithreaded tracker
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) {
+            LongOpenHashSet set = player.moonrise$getChunkLoader().getSentChunksRaw();
+            synchronized (set) {
+                return LongSets.unmodifiable(new LongOpenHashSet(set));
+            }
+        }
+        // Leaf end - Multithreaded tracker
         return LongSets.unmodifiable(player.moonrise$getChunkLoader().getSentChunksRaw().clone()); // Paper - rewrite chunk system
     }
 
@@ -80,10 +88,21 @@ public final class FeatureHooks {
         final LongOpenHashSet rawChunkKeys = player.moonrise$getChunkLoader().getSentChunksRaw();
         final ObjectSet<org.bukkit.Chunk> chunks = new ObjectOpenHashSet<>(rawChunkKeys.size());
         final World world = player.level().getWorld();
-        final LongIterator iter = player.moonrise$getChunkLoader().getSentChunksRaw().longIterator();
-        while (iter.hasNext()) {
-            chunks.add(world.getChunkAt(iter.nextLong(), false));
+        // Leaf start - Multithreaded tracker
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) {
+            synchronized (rawChunkKeys) {
+                final LongIterator iter = rawChunkKeys.longIterator();
+                while (iter.hasNext()) {
+                    chunks.add(world.getChunkAt(iter.nextLong(), false));
+                }
+            }
+        } else {
+            final LongIterator iter = rawChunkKeys.longIterator();
+            while (iter.hasNext()) {
+                chunks.add(world.getChunkAt(iter.nextLong(), false));
+            }
         }
+        // Leaf end - Multithreaded tracker
         // Paper end - rewrite chunk system
         return ObjectSets.unmodifiable(chunks);
     }
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 2e18112a07ee3a9fc5ded8cd28c91ad7cb64ad3f..6c766031524d3c0c7a30f1aa1a0d2bfaf9d482c7 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1956,6 +1956,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Leaf end - SparklyPaper - parallel world ticking
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) { for (ServerLevel world : getAllLevels()) { world.leaf$asyncTracker.onTickEnd(); } } // Leaf - Multithreaded tracker
+
         profiler.popPush("connection");
         this.tickConnection();
         profiler.popPush("players");
@@ -2047,6 +2049,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
         newLevels.remove(level.dimension());
         if (org.dreeam.leaf.config.modules.async.SparklyPaperParallelWorldTicking.enabled) level.tickExecutor.shutdown(); // Leaf - SparklyPaper - parallel world ticking (We remove it in here instead of ServerLevel.close() because ServerLevel.close() is never called!)
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) { level.leaf$asyncTracker.onTickEnd(); } // Leaf - Multithreaded tracker
         this.levels = Collections.unmodifiableMap(newLevels);
     }
     // CraftBukkit end
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 63fb24619ff6f0399c7da9110a2e94fb6b72d390..bd91d9ae11595dcfcbca22cd5a62c9fc6e40e077 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1096,6 +1096,12 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
     // Paper end - optimise entity tracker
 
     protected void tick() {
+        // Leaf start - Multithreaded tracker
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) {
+            this.level.leaf$asyncTracker.tick(this.level);
+            return;
+        }
+        // Leaf end - Multithreaded tracker
         // Paper start - optimise entity tracker
         if (true) {
             this.newTrackerTick();
@@ -1242,12 +1248,61 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         private final Entity entity;
         private final int range;
         private SectionPos lastSectionPos;
-        public final Set<ServerPlayerConnection> seenBy = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(); // Paper - Perf: optimise map impl
+        // Leaf start - Multithreaded tracker
+        public static final ServerPlayerConnection[] EMPTY_OBJECT_ARRAY = new ServerPlayerConnection[0];
+
+        private class SeenBySet extends it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<ServerPlayerConnection> {
+            @Override
+            public boolean add(final ServerPlayerConnection serverPlayerConnection) {
+                if (super.add(serverPlayerConnection)) {
+                    // for plugin compatibility
+                    TrackedEntity.this.seenByUpdated = true;
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+
+            @Override
+            public boolean remove(final Object k) {
+                if (super.remove(k)) {
+                    // for plugin compatibility
+                    TrackedEntity.this.seenByUpdated = true;
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+
+            @Override
+            public void clear() {
+                TrackedEntity.this.seenByUpdated = true;
+                super.clear();
+            }
+        }
+
+        public final Set<ServerPlayerConnection> seenBy = org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled ? it.unimi.dsi.fastutil.objects.ReferenceSets.synchronize(new SeenBySet()) : new SeenBySet(); // Paper - Perf: optimise map impl
+        private volatile boolean seenByUpdated = false;
+        private volatile ServerPlayerConnection[] seenByArray = EMPTY_OBJECT_ARRAY;
+        // Used for iteration, copy on updated
+        public ServerPlayerConnection[] seenBy() {
+            if (!seenByUpdated) {
+                return seenByArray;
+            } else {
+                return seenBy.toArray(EMPTY_OBJECT_ARRAY);
+            }
+        }
+        public void seenByUpdated() {
+            this.seenByArray = this.seenBy.toArray(EMPTY_OBJECT_ARRAY);
+            seenByUpdated = false;
+        }
+        // Leaf end - Multithreaded tracker
 
         // Paper start - optimise entity tracker
         private long lastChunkUpdate = -1L;
         private ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk lastTrackedChunk;
 
+        // Leaf - Multithreaded tracker - diff on change
         @Override
         public final void moonrise$tick(final ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk chunk) {
             if (chunk == null) {
@@ -1257,6 +1312,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
             final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> players = chunk.getPlayers(ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.VIEW_DISTANCE);
 
+            // Leaf - Multithreaded tracker - diff on change
             if (players == null) {
                 this.moonrise$clearPlayers();
                 return;
@@ -1269,27 +1325,35 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             this.lastTrackedChunk = chunk;
 
             final ServerPlayer[] playersRaw = players.getRawDataUnchecked();
-
-            for (int i = 0, len = players.size(); i < len; ++i) {
+            // Leaf start - Multithreaded tracker
+            final int playersLength = Math.min(playersRaw.length, players.size());
+            boolean updated = false;
+            for (int i = 0; i < playersLength; ++i) {
                 final ServerPlayer player = playersRaw[i];
-                this.updatePlayer(player);
+                updated |= this.updatePlayerMulti(player);
             }
+            if (updated) seenByUpdated();
+            // Leaf end - Multithreaded tracker
 
             if (lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk) {
                 // need to purge any players possible not in the chunk list
-                for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+                // Leaf start - Multithreaded tracker
+                boolean removed = false;
+                for (final ServerPlayerConnection conn : this.seenBy()) {
                     final ServerPlayer player = conn.getPlayer();
                     if (!players.contains(player)) {
-                        this.removePlayer(player);
+                        removed |= this.removePlayerMulti(player);
                     }
                 }
+                if (removed) this.seenByUpdated();
+                // Leaf end - Multithreaded tracker
             }
         }
 
         @Override
         public final void moonrise$removeNonTickThreadPlayers() {
             boolean foundToRemove = false;
-            for (final ServerPlayerConnection conn : this.seenBy) {
+            for (final ServerPlayerConnection conn : this.seenBy()) { // Leaf - Multithreaded tracker
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(conn.getPlayer())) {
                     foundToRemove = true;
                     break;
@@ -1300,12 +1364,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 return;
             }
 
-            for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+            for (final ServerPlayerConnection conn : this.seenBy()) { // Leaf - Multithreaded tracker
                 ServerPlayer player = conn.getPlayer();
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
-                    this.removePlayer(player);
+                    this.removePlayerMulti(player); // Leaf - Multithreaded tracker
                 }
             }
+            this.seenByUpdated(); // Leaf - Multithreaded tracker
         }
 
         @Override
@@ -1315,10 +1380,11 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             if (this.seenBy.isEmpty()) {
                 return;
             }
-            for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+            for (final ServerPlayerConnection conn : this.seenBy()) { // Leaf - Multithreaded tracker
                 ServerPlayer player = conn.getPlayer();
-                this.removePlayer(player);
+                this.removePlayerMulti(player); // Leaf - Multithreaded tracker
             }
+            this.seenByUpdated(); // Leaf - Multithreaded tracker
         }
 
         @Override
@@ -1346,7 +1412,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayers(final Packet<? super ClientGamePacketListener> packet) {
-            for (ServerPlayerConnection connection : this.seenBy) {
+            for (ServerPlayerConnection connection : this.seenBy()) { // Leaf - petal - Multithreaded tracker
                 connection.send(packet);
             }
         }
@@ -1361,7 +1427,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayersFiltered(final Packet<? super ClientGamePacketListener> packet, final Predicate<ServerPlayer> targetPredicate) {
-            for (ServerPlayerConnection connection : this.seenBy) {
+            for (ServerPlayerConnection connection : this.seenBy()) { // Leaf - petal - Multithreaded tracker
                 if (targetPredicate.test(connection.getPlayer())) {
                     connection.send(packet);
                 }
@@ -1369,7 +1435,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         }
 
         public void broadcastRemoved() {
-            for (ServerPlayerConnection connection : this.seenBy) {
+            for (ServerPlayerConnection connection : this.seenBy()) { // Leaf - petal - Multithreaded tracker
                 this.serverEntity.removePairing(connection.getPlayer());
             }
         }
@@ -1378,6 +1444,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             org.spigotmc.AsyncCatcher.catchOp("player tracker clear"); // Spigot
             if (this.seenBy.remove(player.connection)) {
                 this.serverEntity.removePairing(player);
+                this.seenByUpdated(); // Leaf - Multithreaded tracker
                 if (this.seenBy.isEmpty()) {
                     ChunkMap.this.level.debugSynchronizers().dropEntity(this.entity);
                 }
@@ -1416,6 +1483,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 // CraftBukkit end
                 if (visibleToPlayer) {
                     if (this.seenBy.add(player.connection)) {
+                        this.seenByUpdated(); // Leaf - Multithreaded tracker
                         // Paper start - entity tracking events
                         if (io.papermc.paper.event.player.PlayerTrackEntityEvent.getHandlerList().getRegisteredListeners().length == 0 || new io.papermc.paper.event.player.PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity()).callEvent()) {
                         this.serverEntity.addPairing(player);
@@ -1429,7 +1497,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                         this.serverEntity.onPlayerAdd(); // Paper - fix desync when a player is added to the tracker
                     }
                 } else {
-                    this.removePlayer(player);
+                    this.removePlayer(player); // Leaf - Multithreaded tracker - diff on change
                 }
             }
         }
@@ -1459,10 +1527,137 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             // Paper end - optimise entity tracker
         }
 
-        public void updatePlayers(final List<ServerPlayer> players) {
+        // Leaf start - Multithreaded tracker
+        public final boolean leaf$tick(final org.dreeam.leaf.async.tracker.TrackerCtx ctx, final ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk chunk) {
+            if (chunk == null || chunk.playersTracking.isEmpty()) {
+                this.lastChunkUpdate = -1L;
+                this.lastTrackedChunk = null;
+                for (final ServerPlayerConnection conn : this.seenBy()) {
+                    ctx.collectStopSeenByPlayer(conn, this);
+                }
+                return false;
+            }
+
+            final it.unimi.dsi.fastutil.objects.ReferenceSet<ServerPlayer> players = chunk.playersTracking;
+            final long currChunkUpdate = chunk.getAtomicUpdateCount();
+            final boolean chunkStateChanged = this.lastChunkUpdate != currChunkUpdate || this.lastTrackedChunk != chunk;
+            this.lastChunkUpdate = currChunkUpdate;
+            this.lastTrackedChunk = chunk;
+
+            final double ex = this.entity.getX();
+            final double ey = this.entity.getY();
+            final double ez = this.entity.getZ();
+            final int eChunkX = this.entity.chunkPosition().x();
+            final int eChunkZ = this.entity.chunkPosition().z();
+            final double effectiveRange = this.getEffectiveRange();
+            final double rangeSqr = effectiveRange * effectiveRange;
+            final double rangeY = level.paperConfig().entities.trackingRangeY.enabled ? level.paperConfig().entities.trackingRangeY.get(this.entity, -1) : -1;
+            final double rangeYSqr = (rangeY > 0.0) ? (rangeY * rangeY) : 0.0;
+            synchronized (players) {
+                for (ServerPlayer player : players) {
+                    if (player == this.entity) {
+                        continue;
+                    }
+                    Vec3 playerPos = player.position();
+                    final double dx = playerPos.x - ex;
+                    final double dz = playerPos.z - ez;
+                    final double dy = playerPos.y - ey;
+                    final double playerViewDistance = ChunkMap.this.getPlayerViewDistance(player);
+                    final boolean flag = ((dx * dx + dz * dz) <= Math.min(rangeSqr, playerViewDistance * playerViewDistance * 256.0))
+                        && ((rangeYSqr == 0.0) || ((dy * dy) <= rangeYSqr))
+                        && this.entity.broadcastToPlayer(player)
+                        && ChunkMap.this.isChunkTracked(player, eChunkX, eChunkZ)
+                        && player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity());
+                    if (flag) {
+                        if (!this.seenBy.contains(player.connection)) {
+                            ctx.collectStartSeenByPlayer(player.connection, this);
+                        }
+                    } else if (this.seenBy.contains(player.connection)) {
+                        ctx.collectStopSeenByPlayer(player.connection, this);
+                    }
+                }
+            }
+            if (!chunkStateChanged) {
+                return this.seenBy().length != 0;
+            }
+            for (final ServerPlayerConnection conn : this.seenBy()) {
+                final ServerPlayer player = conn.getPlayer();
+                if (!players.contains(player)) {
+                    ctx.collectStopSeenByPlayer(conn, this);
+                }
+            }
+            return this.seenBy().length != 0;
+        }
+
+        private boolean removePlayerMulti(final ServerPlayer player) {
+            if (this.seenBy.remove(player.connection)) {
+                this.serverEntity.removePairing(player);
+                if (this.seenBy.isEmpty()) {
+                    ChunkMap.this.level.debugSynchronizers().dropEntity(this.entity);
+                }
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        public boolean updatePlayerMulti(final ServerPlayer player) {
+            org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
+            if (player != this.entity) {
+                // Paper start - remove allocation of Vec3D here
+                // Vec3 deltaToPlayer = player.position().subtract(this.entity.position());
+                double deltaToPlayerX = player.getX() - this.entity.getX();
+                double deltaToPlayerZ = player.getZ() - this.entity.getZ();
+                // Paper end - remove allocation of Vec3D here
+                int playerViewDistance = ChunkMap.this.getPlayerViewDistance(player);
+                double visibleRange = Math.min(this.getEffectiveRange(), playerViewDistance * 16);
+                double distanceSquared = deltaToPlayerX * deltaToPlayerX + deltaToPlayerZ * deltaToPlayerZ; // Paper
+                double rangeSquared = visibleRange * visibleRange;
+                // Paper start - Configurable entity tracking range by Y
+                boolean visibleToPlayer = distanceSquared <= rangeSquared;
+                if (visibleToPlayer && level.paperConfig().entities.trackingRangeY.enabled) {
+                    double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
+                    if (rangeY != -1) {
+                        double deltaToPlayerY = player.getY() - this.entity.getY();
+                        visibleToPlayer = deltaToPlayerY * deltaToPlayerY <= rangeY * rangeY;
+                    }
+                }
+                visibleToPlayer = visibleToPlayer && this.entity.broadcastToPlayer(player) && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x(), this.entity.chunkPosition().z());
+                // Paper end - Configurable entity tracking range by Y
+                // CraftBukkit start - respect vanish API
+                if (visibleToPlayer && !player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity())) { // Paper - only consider hits // SparklyPaper - optimize canSee checks
+                    visibleToPlayer = false;
+                }
+                // CraftBukkit end
+                if (visibleToPlayer) {
+                    if (this.seenBy.add(player.connection)) {
+                        // Paper start - entity tracking events
+                        if (io.papermc.paper.event.player.PlayerTrackEntityEvent.getHandlerList().getRegisteredListeners().length == 0 || new io.papermc.paper.event.player.PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity()).callEvent()) {
+                            this.serverEntity.addPairing(player);
+                            if (this.seenBy.size() == 1) {
+                                ChunkMap.this.level.debugSynchronizers().registerEntity(this.entity);
+                            }
+
+                            ChunkMap.this.level.debugSynchronizers().startTrackingEntity(player, this.entity);
+                        }
+                        // Paper end - entity tracking events
+                        this.serverEntity.onPlayerAdd(); // Paper - fix desync when a player is added to the tracker
+                        return true;
+                    }
+                } else {
+                    this.removePlayerMulti(player);
+                }
+            }
+            return false;
+        }
+
+        private void updatePlayers(final List<ServerPlayer> players) {
+            boolean updated = false;
             for (ServerPlayer player : players) {
-                this.updatePlayer(player);
+                updated |= this.updatePlayerMulti(player);
             }
+            if (updated) seenByUpdated();
         }
+        // Leaf end - Multithreaded tracker
     }
 }
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 4f2344b6e4fe6aebdbb7ca2ad65365515d8e2424..e4bac62e980ea52c42b2345e9b486e10c499c85f 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -60,11 +60,11 @@ public class ServerEntity {
     public static final int FORCED_POS_UPDATE_PERIOD = 60;
     private static final int FORCED_TELEPORT_PERIOD = 400;
     private final ServerLevel level;
-    private final Entity entity;
+    public final Entity entity; // Leaf - Multithreaded tracker - private -> public
     private final UpdateInterval updateInterval;
     private final boolean trackDelta;
     private final ServerEntity.Synchronizer synchronizer;
-    private final InterpolationTracker interpolationTracker;
+    public final InterpolationTracker interpolationTracker;
     private final VecDeltaCodec positionCodec = new VecDeltaCodec();
     private byte lastSentYRot;
     private byte lastSentXRot;
@@ -75,7 +75,7 @@ public class ServerEntity {
     private List<Entity> lastPassengers = com.google.common.collect.ImmutableList.of(); // Paper - optimize passenger checks
     private boolean wasRiding;
     private boolean wasOnGround;
-    private @Nullable List<SynchedEntityData.DataValue<?>> trackedDataValues;
+    public @Nullable List<SynchedEntityData.DataValue<?>> trackedDataValues; // Leaf - Multithreaded tracker - private -> public
     private final Set<net.minecraft.server.network.ServerPlayerConnection> trackedPlayers; // Paper
 
     public ServerEntity(
@@ -88,7 +88,7 @@ public class ServerEntity {
     ) {
         this.trackedPlayers = trackedPlayers; // Paper
         this.level = level;
-        this.synchronizer = synchronizer;
+        this.synchronizer = synchronizer; // Leaf - Multithreaded tracker - diff on change
         this.entity = entity;
         this.updateInterval = updateInterval;
         this.trackDelta = trackDelta;
@@ -103,20 +103,21 @@ public class ServerEntity {
     }
 
     // Paper start - fix desync when a player is added to the tracker
-    private boolean forceStateResync;
+    private boolean forceStateResync; // Leaf - Multithreaded tracker - diff on change
     public void onPlayerAdd() {
         this.forceStateResync = true;
     }
     // Paper end - fix desync when a player is added to the tracker
 
-    public void sendChanges() {
+    public void sendChanges() { // Leaf - Multithreaded tracker - diff on change
         // Paper start - optimise collisions
         if (((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this.entity).moonrise$isHardColliding()) {
             this.teleportDelay = 9999;
         }
+        // Leaf - Multithreaded tracker - diff on change
         // Paper end - optimise collisions
-        this.entity.updateDataBeforeSync();
-        List<Entity> passengers = this.entity.getPassengers();
+        this.entity.updateDataBeforeSync(); // Leaf - Multithreaded tracker - diff on change
+        List<Entity> passengers = this.entity.getPassengers(); // Leaf - Multithreaded tracker - diff on change
         if (!passengers.equals(this.lastPassengers)) {
             this.synchronizer.sendToTrackingPlayers(new ClientboundSetPassengersPacket(this.entity));
             // Paper start - Allow riding players
@@ -124,27 +125,27 @@ public class ServerEntity {
                 player.connection.send(new ClientboundSetPassengersPacket(this.entity));
             }
             // Paper end - Allow riding players
-            this.lastPassengers = passengers;
+            this.lastPassengers = passengers; // Leaf - Multithreaded tracker - diff on change
         }
 
-        if (!this.trackedPlayers.isEmpty() && this.entity instanceof ItemFrame frame /*&& this.tickCount % 10 == 0*/) { // CraftBukkit - moved tickCount below // Paper - Perf: Only tick item frames if players can see it
+        if (!this.trackedPlayers.isEmpty() && this.entity instanceof ItemFrame frame /*&& this.tickCount % 10 == 0*/) { // CraftBukkit - moved tickCount below // Paper - Perf: Only tick item frames if players can see it // Leaf - Multithreaded tracker - diff on change
             ItemStack itemStack = frame.getItem();
             if (this.level.paperConfig().maps.itemFrameCursorUpdateInterval > 0 && this.tickCount % this.level.paperConfig().maps.itemFrameCursorUpdateInterval == 0 && itemStack.getItem() instanceof MapItem) { // CraftBukkit - Moved this.tickCounter % 10 logic here so item frames do not enter the other blocks // Paper - Make item frame map cursor update interval configurable
                 MapId id = frame.cachedMapId; // Paper - Perf: Cache map ids on item frames
-                MapItemSavedData data = MapItem.getSavedData(id, this.level);
+                MapItemSavedData data = MapItem.getSavedData(id, this.level); // Leaf - Multithreaded tracker - diff on change
                 if (data != null) {
                     for (final net.minecraft.server.network.ServerPlayerConnection connection : this.trackedPlayers) { // Paper
                         final ServerPlayer player = connection.getPlayer(); // Paper
                         data.tickCarriedBy(player, itemStack, frame);
-                        Packet<?> packet = data.getUpdatePacket(id, player);
+                        Packet<?> packet = data.getUpdatePacket(id, player); // Leaf - Multithreaded tracker - diff on change
                         if (packet != null) {
-                            player.connection.send(packet);
+                            player.connection.send(packet); // Leaf - Multithreaded tracker - diff on change
                         }
                     }
                 }
             }
 
-            this.sendDirtyEntityData();
+            this.sendDirtyEntityData(); // Leaf - Multithreaded tracker - diff on change
         }
 
         Vec3 currentPosition = this.entity.trackingPosition();
@@ -157,27 +158,27 @@ public class ServerEntity {
         if (this.forceStateResync || this.entity.needsSync || this.updateInterval.test(this.tickCount) || this.entity.getEntityData().isDirty()) { // Paper - fix desync when a player is added to the tracker
             byte yRotn = Mth.packDegrees(this.entity.getYRot());
             byte xRotn = Mth.packDegrees(this.entity.getXRot());
-            boolean shouldSendRotation = Math.abs(yRotn - this.lastSentYRot) >= 1 || Math.abs(xRotn - this.lastSentXRot) >= 1;
+            boolean shouldSendRotation = Math.abs(yRotn - this.lastSentYRot) >= 1 || Math.abs(xRotn - this.lastSentXRot) >= 1; // Leaf - Multithreaded tracker - diff on change
             if (this.entity.isPassenger()) {
-                if (shouldSendRotation) {
+                if (shouldSendRotation) { // Leaf - Multithreaded tracker - diff on change
                     this.synchronizer.sendToTrackingPlayers(new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround()));
                     this.lastSentYRot = yRotn;
                     this.lastSentXRot = xRotn;
-                }
+                } // Leaf - Multithreaded tracker - diff on change
 
                 this.interpolationTracker.clear();
                 this.positionCodec.setBase(currentPosition);
                 this.sendDirtyEntityData();
                 this.wasRiding = true;
             } else if (this.entity instanceof AbstractMinecart minecart && minecart.getBehavior() instanceof NewMinecartBehavior newMinecartBehavior) {
-                this.handleMinecartPosRot(newMinecartBehavior, yRotn, xRotn, shouldSendRotation);
+                this.handleMinecartPosRot(newMinecartBehavior, yRotn, xRotn, shouldSendRotation); // Leaf - Multithreaded tracker - diff on change
             } else {
                 this.teleportDelay++;
                 PositionPath position = this.interpolationTracker.getPositionPath(currentPosition);
                 this.interpolationTracker.clear();
                 MovementPacket<ClientGamePacketListener> packet;
                 if (this.entity instanceof AbstractArrow) {
-                    packet = this.createMovePacket(position, yRotn, xRotn, true, true);
+                    packet = this.createMovePacket(position, yRotn, xRotn, true, true, this.forceStateResync); // Leaf - Multithreaded tracker
                 } else {
                     // Paper start - reduce allocation of Vec3 here
                     Vec3 base = this.positionCodec.base;
@@ -187,7 +188,7 @@ public class ServerEntity {
                     boolean positionChanged = (dx * dx + dy * dy + dz * dz) >= 7.62939453125E-6D;
                     // Paper end - reduce allocation of Vec3 here
                     boolean shouldSendPosition = positionChanged || this.tickCount % 60 == 0;
-                    packet = this.createMovePacket(position, yRotn, xRotn, shouldSendPosition, shouldSendRotation);
+                    packet = this.createMovePacket(position, yRotn, xRotn, shouldSendPosition, shouldSendRotation, this.forceStateResync); // Leaf - Multithreaded tracker
                 }
 
                 if (org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.reduceUselessEntityMovePackets && isUselessMoveEntityPacket(packet, shouldSendRotation)) packet = null; // Leaf - Dont send useless entity packets
@@ -197,25 +198,25 @@ public class ServerEntity {
                     if (movement != this.lastSentMovement) { // SparklyPaper start - skip distanceToSqr call in ServerEntity#sendChanges if the delta movement hasn't changed
                     double diff = movement.distanceToSqr(this.lastSentMovement);
                     if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
-                        this.lastSentMovement = movement;
+                        this.lastSentMovement = movement; // Leaf - Multithreaded tracker - diff on change
                         if (this.entity instanceof AbstractHurtingProjectile projectile) {
                             this.synchronizer
                                 .sendToTrackingPlayers(
                                     new ClientboundBundlePacket(
-                                        List.of(
+                                        List.of( // Leaf - Multithreaded tracker - diff on change
                                             new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement),
                                             new ClientboundProjectilePowerPacket(projectile.getId(), projectile.accelerationPower)
                                         )
                                     )
                                 );
-                        } else {
+                        } else { // Leaf - Multithreaded tracker - diff on change
                             this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                         }
                     }
                     } // SparklyPaper end
                 }
 
-                if (packet != null) {
+                if (packet != null) { // Leaf - Multithreaded tracker - diff on change
                     this.synchronizer.sendToTrackingPlayers(packet);
                     this.sendDirtyEntityData();
                     if (packet.hasPosition()) {
@@ -230,16 +231,16 @@ public class ServerEntity {
                     this.sendDirtyEntityData();
                 }
 
-                this.wasRiding = false;
+                this.wasRiding = false; // Leaf - Multithreaded tracker - diff on change
             }
 
-            byte yHeadRot = Mth.packDegrees(this.entity.getYHeadRot());
+            byte yHeadRot = Mth.packDegrees(this.entity.getYHeadRot()); // Leaf - Multithreaded tracker - diff on change
             if (Math.abs(yHeadRot - this.lastSentYHeadRot) >= 1) {
                 this.synchronizer.sendToTrackingPlayers(new ClientboundRotateHeadPacket(this.entity, yHeadRot));
                 this.lastSentYHeadRot = yHeadRot;
             }
 
-            this.entity.needsSync = false;
+            this.entity.needsSync = false; // Leaf - Multithreaded tracker - diff on change
             this.forceStateResync = false; // Paper - fix desync when a player is added to the tracker
         }
 
@@ -248,21 +249,21 @@ public class ServerEntity {
             // CraftBukkit start - Create PlayerVelocity event
             boolean cancelled = false;
 
-            if (this.entity instanceof ServerPlayer) {
+            if (this.entity instanceof ServerPlayer) { // Leaf - Multithreaded tracker - diff on change
                 org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.entity.getBukkitEntity();
                 org.bukkit.util.Vector velocity = player.getVelocity();
 
-                org.bukkit.event.player.PlayerVelocityEvent event = new org.bukkit.event.player.PlayerVelocityEvent(player, velocity.clone());
+                org.bukkit.event.player.PlayerVelocityEvent event = new org.bukkit.event.player.PlayerVelocityEvent(player, velocity.clone()); // Leaf - Multithreaded tracker - diff on change
                 if (!event.callEvent()) {
-                    cancelled = true;
+                    cancelled = true; // Leaf - Multithreaded tracker - diff on change
                 } else if (!velocity.equals(event.getVelocity())) {
                     player.setVelocity(event.getVelocity());
-                }
+                } // Leaf - Multithreaded tracker - diff on change
             }
 
             if (cancelled) {
                 return;
-            }
+            } // Leaf - Multithreaded tracker - diff on change
             // CraftBukkit end
             this.entity.syncVelocity = false;
             this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundSetEntityMotionPacket(this.entity));
@@ -270,9 +271,9 @@ public class ServerEntity {
     }
 
     private @Nullable MovementPacket<ClientGamePacketListener> createMovePacket(
-        final PositionPath position, final byte yRotn, final byte xRotn, final boolean shouldSendPosition, final boolean shouldSendRotation
+        final PositionPath position, final byte yRotn, final byte xRotn, final boolean shouldSendPosition, final boolean shouldSendRotation, final boolean force // Leaf - Multithreaded tracker
     ) {
-        if (this.forceStateResync || this.entity.getRequiresPrecisePosition() || this.teleportDelay > 400 || this.wasRiding || this.wasOnGround != this.entity.onGround()) { // Paper - fix desync when a player is added to the tracker
+        if (force || this.entity.getRequiresPrecisePosition() || this.teleportDelay > 400 || this.wasRiding || this.wasOnGround != this.entity.onGround()) { // Paper - fix desync when a player is added to the tracker // Leaf - Multithreaded tracker
             this.wasOnGround = this.entity.onGround();
             this.teleportDelay = 0;
             return ClientboundEntityPositionSyncPacket.of(this.entity, position);
@@ -328,14 +329,14 @@ public class ServerEntity {
     private void handleMinecartPosRot(final NewMinecartBehavior newMinecartBehavior, final byte yRotn, final byte xRotn, final boolean shouldSendRotation) {
         this.sendDirtyEntityData();
         if (newMinecartBehavior.lerpSteps.isEmpty()) {
-            Vec3 movement = this.entity.getDeltaMovement();
+            Vec3 movement = this.entity.getDeltaMovement(); // Leaf - Multithreaded tracker - diff on change
             double diff = movement.distanceToSqr(this.lastSentMovement);
             Vec3 currentPosition = this.entity.trackingPosition();
             boolean positionChanged = this.positionCodec.delta(currentPosition).lengthSqr() >= 7.6293945E-6F;
             boolean shouldSendPosition = positionChanged || this.tickCount % 60 == 0;
             if (shouldSendPosition || shouldSendRotation || diff > 1.0E-7) {
                 this.synchronizer
-                    .sendToTrackingPlayers(
+                    .sendToTrackingPlayers( // Leaf - Multithreaded tracker - diff on change
                         new ClientboundMoveMinecartPacket(
                             this.entity.getId(),
                             List.of(
@@ -444,6 +445,7 @@ public class ServerEntity {
         return Mth.unpackDegrees(this.lastSentYHeadRot);
     }
 
+    // Leaf - Multithreaded tracker - diff on change
     private void sendDirtyEntityData() {
         SynchedEntityData entityData = this.entity.getEntityData();
         List<SynchedEntityData.DataValue<?>> packedValues = entityData.packDirty();
@@ -451,10 +453,12 @@ public class ServerEntity {
             this.trackedDataValues = entityData.getNonDefaultValues();
             this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundSetEntityDataPacket(this.entity.getId(), packedValues));
         }
+        // Leaf - Multithreaded tracker - diff on change
 
         if (this.entity instanceof LivingEntity livingEntity) {
             Set<AttributeInstance> attributes = livingEntity.getAttributes().getAttributesToSync();
             if (!attributes.isEmpty()) {
+                // Leaf - Multithreaded tracker - diff on change
                 // CraftBukkit start - Send scaled max health
                 if (this.entity instanceof ServerPlayer serverPlayer) {
                     serverPlayer.getBukkitEntity().injectScaledMaxHealth(attributes, false);
@@ -467,6 +471,170 @@ public class ServerEntity {
         }
     }
 
+    // Leaf start - Multithreaded tracker
+    public void leaf$onTrackingStart(final ChunkMap.TrackedEntity tracker) {
+        this.level.leaf$asyncTracker.addTracker(tracker);
+    }
+
+    public void leaf$onTrackingEnd(final ChunkMap.TrackedEntity tracker) {
+        this.level.leaf$asyncTracker.removeTracker(tracker);
+    }
+
+    public void leaf$sendChanges(final org.dreeam.leaf.async.tracker.TrackerCtx ctx, final ChunkMap.TrackedEntity tracker, org.dreeam.leaf.async.tracker.TrackerInput input, final boolean force) {
+        if (this.forceStateResync && !force) {
+            this.forceStateResync = false;
+            ctx.forceResync(tracker, input);
+            return;
+        }
+        if (this.entity.leaf$wantUpdateData()) {
+            ctx.wantUpdateData(tracker);
+        }
+        // Paper start - optimise collisions
+        if (this.entity.moonrise$isHardColliding()) {
+            this.teleportDelay = 9999;
+        }
+        // Paper end - optimise collisions
+        //this.entity.updateDataBeforeSync(); // Move to AsyncTracker#handlePlayer
+        List<Entity> passengers = this.entity.getPassengers();
+        if (!passengers.equals(this.lastPassengers)) {
+            ctx.sendToTrackingPlayers(tracker, new ClientboundSetPassengersPacket(this.entity));
+            // Paper start - Allow riding players
+            if (this.entity instanceof ServerPlayer player) {
+                ctx.send(player.connection, new ClientboundSetPassengersPacket(this.entity));
+            }
+            // Paper end - Allow riding players
+            this.lastPassengers = passengers;
+        }
+        if (!tracker.seenBy.isEmpty() && this.entity instanceof ItemFrame frame /*&& this.tickCount % 10 == 0*/) { // CraftBukkit - moved tickCount below // Paper - Perf: Only tick item frames if players can see it
+            if (this.level.paperConfig().maps.itemFrameCursorUpdateInterval > 0 && this.tickCount % this.level.paperConfig().maps.itemFrameCursorUpdateInterval == 0 && frame.getItem().getItem() instanceof MapItem) { // CraftBukkit - Moved this.tickCounter % 10 logic here so item frames do not enter the other blocks // Paper - Make item frame map cursor update interval configurable
+                ctx.updateItemFrame(frame);
+            }
+            ctx.sendDirtyEntityData(tracker);
+        }
+
+        Vec3 currentPosition = input.position;
+        this.interpolationTracker.updateTracking(currentPosition);
+        if (input.syncPosition) {
+            this.tickCount = this.updateInterval.nextInterval(this.tickCount);
+            input.syncPosition = false;
+        }
+        if (force || this.entity.needsSync || this.updateInterval.test(this.tickCount) || this.entity.getEntityData().isDirty()) { // Paper - fix desync when a player is added to the tracker
+            byte yRotn = Mth.packDegrees(this.entity.getYRot());
+            byte xRotn = Mth.packDegrees(this.entity.getXRot());
+            boolean shouldSendRotation = Math.abs(yRotn - this.lastSentYRot) >= 1 || Math.abs(xRotn - this.lastSentXRot) >= 1;
+            if (this.entity.isPassenger()) {
+                if (shouldSendRotation) {
+                    ctx.sendToTrackingPlayers(tracker, new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround()));
+                    this.lastSentYRot = yRotn;
+                    this.lastSentXRot = xRotn;
+                }
+
+                this.interpolationTracker.clear();
+                this.positionCodec.setBase(currentPosition);
+                ctx.sendDirtyEntityData(tracker);
+                this.wasRiding = true;
+            } else if (this.entity instanceof AbstractMinecart minecart && minecart.getBehavior() instanceof NewMinecartBehavior newMinecartBehavior) {
+                this.leaf$handleMinecartPosRot(ctx, currentPosition, tracker, newMinecartBehavior, yRotn, xRotn, shouldSendRotation);
+            } else {
+                this.teleportDelay++;
+                PositionPath position = this.interpolationTracker.getPositionPath(currentPosition);
+                this.interpolationTracker.clear();
+                MovementPacket<ClientGamePacketListener> packet;
+                if (this.entity instanceof AbstractArrow) {
+                    packet = this.createMovePacket(position, yRotn, xRotn, true, true, force);
+                } else {
+                    // Paper start - reduce allocation of Vec3 here
+                    Vec3 base = this.positionCodec.base;
+                    double dx = currentPosition.x - base.x;
+                    double dy = currentPosition.y - base.y;
+                    double dz = currentPosition.z - base.z;
+                    boolean positionChanged = (dx * dx + dy * dy + dz * dz) >= 7.62939453125E-6D;
+                    // Paper end - reduce allocation of Vec3 here
+                    boolean shouldSendPosition = positionChanged || this.tickCount % 60 == 0;
+                    packet = this.createMovePacket(position, yRotn, xRotn, shouldSendPosition, shouldSendRotation, force);
+                }
+
+                if (org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.reduceUselessEntityMovePackets && isUselessMoveEntityPacket(packet, shouldSendRotation)) packet = null; // Leaf - Dont send useless entity packets
+
+                if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) {
+                    Vec3 movement = this.entity.getDeltaMovement();
+                    if (movement != this.lastSentMovement) { // SparklyPaper start - skip distanceToSqr call in ServerEntity#sendChanges if the delta movement hasn't changed
+                        double diff = movement.distanceToSqr(this.lastSentMovement);
+                        if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
+                            this.lastSentMovement = movement;
+                            if (this.entity instanceof AbstractHurtingProjectile projectile) {
+                                ctx.sendToTrackingPlayers(tracker, new ClientboundBundlePacket(List.of(
+                                    new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement),
+                                    new ClientboundProjectilePowerPacket(projectile.getId(), projectile.accelerationPower)
+                                )));
+                            } else {
+                                ctx.sendToTrackingPlayers(tracker, new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            }
+                        }
+                    } // SparklyPaper end
+                }
+
+                if (packet != null) {
+                    ctx.sendToTrackingPlayers(tracker, packet);
+                    ctx.sendDirtyEntityData(tracker);
+                    if (packet.hasPosition()) {
+                        this.positionCodec.setBase(currentPosition);
+                    }
+
+                    if (packet.hasRotation()) {
+                        this.lastSentYRot = yRotn;
+                        this.lastSentXRot = xRotn;
+                    }
+                } else {
+                    ctx.sendDirtyEntityData(tracker);
+                }
+
+                this.wasRiding = false;
+            }
+
+            byte yHeadRot = Mth.packDegrees(this.entity.getYHeadRot());
+            if (Math.abs(yHeadRot - this.lastSentYHeadRot) >= 1) {
+                ctx.sendToTrackingPlayers(tracker, new ClientboundRotateHeadPacket(this.entity, yHeadRot));
+                this.lastSentYHeadRot = yHeadRot;
+            }
+
+            this.entity.needsSync = false;
+            //this.forceStateResync = false; // Paper - fix desync when a player is added to the tracker // move up
+        }
+
+        this.tickCount++;
+        if (this.entity.syncVelocity && !(this.entity instanceof ServerPlayer)) {
+            this.entity.syncVelocity = false;
+            ctx.sendToTrackingPlayersAndSelf(tracker, new ClientboundSetEntityMotionPacket(this.entity));
+        }
+    }
+
+    private void leaf$handleMinecartPosRot(final org.dreeam.leaf.async.tracker.TrackerCtx ctx, Vec3 currentPosition, final ChunkMap.TrackedEntity tracker, final NewMinecartBehavior newMinecartBehavior, final byte yRotn, final byte xRotn, final boolean shouldSendRotation) {
+        ctx.sendDirtyEntityData(tracker);
+        if (newMinecartBehavior.lerpSteps.isEmpty()) {
+            Vec3 movement = this.entity.getDeltaMovement();
+            double diff = movement.distanceToSqr(this.lastSentMovement);
+            boolean positionChanged = this.positionCodec.delta(currentPosition).lengthSqr() >= 7.6293945E-6F;
+            boolean shouldSendPosition = positionChanged || this.tickCount % 60 == 0;
+            if (shouldSendPosition || shouldSendRotation || diff > 1.0E-7) {
+                ctx.sendToTrackingPlayers(tracker, new ClientboundMoveMinecartPacket(
+                    this.entity.getId(),
+                    List.of(
+                        new NewMinecartBehavior.MinecartStep(
+                            currentPosition, this.entity.getDeltaMovement(), this.entity.getYRot(), this.entity.getXRot(), 1.0F
+                        ))));
+            }
+        } else {
+            ctx.sendToTrackingPlayers(tracker, new ClientboundMoveMinecartPacket(this.entity.getId(), List.copyOf(newMinecartBehavior.lerpSteps)));
+            newMinecartBehavior.lerpSteps.clear();
+        }
+
+        this.lastSentYRot = yRotn;
+        this.lastSentXRot = xRotn;
+        this.positionCodec.setBase(currentPosition);
+    }
+    // Leaf end - Multithreaded tracker
+
     public interface Synchronizer {
         void sendToTrackingPlayers(Packet<? super ClientGamePacketListener> packet);
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 0d600fc5c06caba1a600eac8a279a1b964fb2c04..2fbd50eeb7af191568bdcc5fecc9a064e4d6f704 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -245,6 +245,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public final org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
+    public final org.dreeam.leaf.async.tracker.AsyncTracker leaf$asyncTracker = new org.dreeam.leaf.async.tracker.AsyncTracker(); // Leaf - Multithreaded tracker
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -1135,6 +1136,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.debugSynchronizers.tick(this.server.debugSubscribers());
         profiler.pop();
         this.chunkSource.randomState().garbageCollect();
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) { this.leaf$asyncTracker.onEntitiesTickEnd(); } // Leaf - Multithreaded tracker
     }
 
     @Override
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index af3ac9e5bd57c3af252a931e9e965f7a01cc79ac..b5cebe13d81a00670d2fd5b06a452c3a4f4c5fc3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -550,7 +550,18 @@ public abstract class Entity
 
     @Override
     public final void moonrise$setTrackedEntity(final net.minecraft.server.level.ChunkMap.TrackedEntity trackedEntity) {
+        net.minecraft.server.level.ChunkMap.TrackedEntity previous = this.trackedEntity; // Leaf - Multithreaded tracker
         this.trackedEntity = trackedEntity;
+        // Leaf start - Multithreaded tracker
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) {
+            if (previous != null) {
+                previous.serverEntity.leaf$onTrackingEnd(previous);
+            }
+            if (trackedEntity != null) {
+                trackedEntity.serverEntity.leaf$onTrackingStart(trackedEntity);
+            }
+        }
+        // Leaf end - Multithreaded tracker
     }
 
     private static void collectIndirectPassengers(final List<Entity> into, final List<Entity> from) {
@@ -905,6 +916,11 @@ public abstract class Entity
         }
     }
 
+    // Leaf start - Multithreaded tracker
+    public boolean leaf$wantUpdateData() {
+        return false;
+    }
+    // Leaf end - Multithreaded tracker
     public void updateDataBeforeSync() {
     }
 
@@ -1356,6 +1372,7 @@ public abstract class Entity
     public void recordMovement(final MoverType moverType, final Vec3 movement) {
         if (moverType.isServerAndClientSimulated() || this.getMoveSimulationType() == MoveSimulationType.SERVER_AND_CLIENT) {
             this.interpolationHandler.applyPredictedMovement(movement);
+            // todo
         }
     }
 
@@ -1400,6 +1417,7 @@ public abstract class Entity
             effectState.getBlock().bounceOn(this.level, effectState, this.getBlockPosBelowThatAffectsMyMovement(), this, this.fallDistance);
             this.gameEvent(GameEvent.BOUNCE);
             this.syncPosition = true;
+            // todo
         }
 
         this.setDeltaMovement(movementAfterBounce);
@@ -5365,6 +5383,12 @@ public abstract class Entity
             this.setBoundingBox(this.makeBoundingBox());
         }
         // Paper end - Block invalid positions and bounding box
+        // todo
+        // Leaf start - Multithreaded tracker
+        if (this.leaf$trackingInput != null) {
+            this.leaf$trackingInput.setPosition(this.trackingPosition());
+        }
+        // Leaf end - Multithreaded tracker
     }
 
     public void checkDespawn() {
diff --git a/net/minecraft/world/entity/InterpolationTracker.java b/net/minecraft/world/entity/InterpolationTracker.java
index f318aee9ec9a6cf1b1a8292f82d095a200f88683..55d9a43dec9162aae001ef5bf4615850f15d83d9 100644
--- a/net/minecraft/world/entity/InterpolationTracker.java
+++ b/net/minecraft/world/entity/InterpolationTracker.java
@@ -11,6 +11,15 @@ public interface InterpolationTracker {
 
     void clear();
 
+    // Leaf start - Multithreaded tracker
+    default void leaf$updateTracking(final Vec3 trackingPos, final org.dreeam.leaf.async.tracker.TrackerInput input) {
+    }
+    // Leaf end - Multithreaded tracker
+
+    default Vec3 leaf$predictedDelta() {
+        return Vec3.ZERO;
+    }
+
     record NoOpInterpolationTracker() implements InterpolationTracker {
         @Override
         public void updateTracking(final Vec3 trackingPos) {
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 16752f51e89592c5658b5d5ec078c5239f43743d..0b0e3a542f882aa5803a07271522b70614bdfb8d 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1011,6 +1011,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         }
     }
 
+    // Leaf start - Multithreaded tracker
+    @Override
+    public boolean leaf$wantUpdateData() {
+        return effectsDirty;
+    }
+    // Leaf end - Multithreaded tracker
+
     private void updateDirtyEffects() {
         if (this.effectsDirty) {
             this.updateInvisibilityStatus();
@@ -2990,6 +2997,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return this.equipment.get(slot);
     }
 
+    // Leaf start - Multithreaded tracker
+    public boolean hasItemInSlots() {
+        return !this.equipment.isEmpty();
+    }
+    // Leaf end - Multithreaded tracker
+
     public void setItemSlot(final EquipmentSlot slot, final ItemStack itemStack) {
         // Paper start
         this.setItemSlot(slot, itemStack, false);
diff --git a/net/minecraft/world/entity/SteppedInterpolationTracker.java b/net/minecraft/world/entity/SteppedInterpolationTracker.java
index cd1ce498f661024cd38f7fc9286a757b69398c38..b7435ef2eaa58c18072bac08b01a51e5bacb8e61 100644
--- a/net/minecraft/world/entity/SteppedInterpolationTracker.java
+++ b/net/minecraft/world/entity/SteppedInterpolationTracker.java
@@ -45,6 +45,27 @@ public class SteppedInterpolationTracker extends InterpolationHandler.NoOpInterp
         }
     }
 
+    // Leaf start - Multithreaded tracker
+    public void leaf$updateTracking(final Vec3 trackingPos, final org.dreeam.leaf.async.tracker.TrackerInput input) {
+        final Vec3 predictedDelta = input.predictedDelta;
+        if (predictedDelta.lengthSqr() > 1.0E-5F) {
+            this.trackedSteps.replaceAll(step -> step.addDelta(predictedDelta));
+        }
+
+        input.predictedDelta = Vec3.ZERO;
+        this.ticksSinceLastStep++;
+        if (input.syncPosition) {
+            input.syncPosition = false;
+            this.addStep(trackingPos);
+        }
+    }
+    // Leaf end - Multithreaded tracker
+
+    @Override
+    public net.minecraft.world.phys.Vec3 leaf$predictedDelta() {
+        return this.predictedDelta;
+    }
+
     @Override
     public PositionPath getPositionPath(final Vec3 currentPosition) {
         this.addStep(currentPosition);
diff --git a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
index 2ed79b97be5a7eb98f3d2da2a643c3d9c7cb0f6f..1ed7bd37ca2b80e715cc9a9d9ea76a7f0c94d49d 100644
--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
@@ -122,14 +122,14 @@ public class WanderingTrader extends AbstractVillager implements Consumable.Over
                     this,
                     PotionContents.createItemStack(Items.POTION, Potions.INVISIBILITY),
                     SoundEvents.WANDERING_TRADER_DISAPPEARED,
-                    e -> this.canDrinkPotion && this.level().isDarkOutside() && !e.isInvisible() // Paper - Add more WanderingTrader API
+                    e -> this.canDrinkPotion && this.level().isDarkOutside() && !e.isInvisible() && !e.hasEffect(net.minecraft.world.effect.MobEffects.INVISIBILITY) // Paper - Add more WanderingTrader API // Leaf - Multithreaded tracker - prevents double use
                 )
             );
         this.goalSelector
             .addGoal(
                 0,
                 new UseItemGoal<>(
-                    this, new ItemStack(Items.MILK_BUCKET), SoundEvents.WANDERING_TRADER_REAPPEARED, e -> level().purpurConfig.milkClearsBeneficialEffects && this.canDrinkMilk && this.level().isBrightOutside() && e.isInvisible() // Paper - Add more WanderingTrader API // Purpur - Milk Keeps Beneficial Effects
+                    this, new ItemStack(Items.MILK_BUCKET), SoundEvents.WANDERING_TRADER_REAPPEARED, e -> level().purpurConfig.milkClearsBeneficialEffects && this.canDrinkMilk && this.level().isBrightOutside() && e.isInvisible() && e.hasEffect(net.minecraft.world.effect.MobEffects.INVISIBILITY) // Paper - Add more WanderingTrader API // Purpur - Milk Keeps Beneficial Effects // Leaf - Multithreaded tracker - prevents double use
                 )
             );
         this.goalSelector.addGoal(1, new TradeWithPlayerGoal(this));
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 6ea8f445bd8d0ab3a061f086e62304f104a82de0..f08def3b40ec970ea61fccdcd44041cbe7baf68a 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -34,13 +34,20 @@ public class NewMinecartBehavior extends MinecartBehavior {
     private int cachedLerpDelay;
     private float cachedPartialTick;
     private int lerpDelay = 0;
-    public final List<NewMinecartBehavior.MinecartStep> lerpSteps = new LinkedList<>();
+    public final List<NewMinecartBehavior.MinecartStep> lerpSteps; // Leaf - Multithreaded tracker
     public final List<NewMinecartBehavior.MinecartStep> currentLerpSteps = new LinkedList<>();
     public double currentLerpStepsTotalWeight = 0.0;
     public NewMinecartBehavior.MinecartStep oldLerp = NewMinecartBehavior.MinecartStep.ZERO;
 
     public NewMinecartBehavior(final AbstractMinecart minecart) {
         super(minecart);
+        // Leaf start - Multithreaded tracker
+        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled) {
+            lerpSteps = it.unimi.dsi.fastutil.objects.ObjectLists.synchronize(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>());
+        } else {
+            lerpSteps = new LinkedList<>();
+        }
+        // Leaf end - Multithreaded tracker
     }
 
     @Override
